pkg/kubelet/images/image_manager.go

```diff
@@ -172,13 +173,14 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
 // applyDefaultImageTag parses a docker image string, if it doesn't contain any tag or digest,
 // a default tag will be applied.
 func applyDefaultImageTag(image string) (string, error) {
-	named, err := dockerref.ParseNormalizedNamed(image)
+	_, tag, _, err := parsers.ParseImageName(image)
 	if err != nil {
 		return "", fmt.Errorf("couldn't parse image reference %q: %v", image, err)
 	}
-	_, isTagged := named.(dockerref.Tagged)
-	_, isDigested := named.(dockerref.Digested)
-	if !isTagged && !isDigested {
+	if strings.Contains(image, ":latest") {
```

Comment:
Should we use `strings.HasSuffix()`  instead?

Reply:
done.

Reply:
I am still seeing `strings.Contains` here. Did you update the PR?

Reply:
Also, can `if strings.HasSuffix(image, ":latest")` be moved to the beginning of this function, to avoid parsing if possible?

Reply:
@liggitt pointed out here that would break cases where we have both a tag and digest: https://github.com/kubernetes/kubernetes/pull/116231#discussion_r1126549268

So I added a test case that verifies this and I had to use strings.Contains for that reason.

Reply:
gotcha, sorry didnt see the comment. and can it be moved to the beginning of this func?

Reply:
not really though.  

```
	ref, err := kubecontainer.GenerateContainerRef(pod, container)
	if err != nil {
		klog.ErrorS(err, "Couldn't make a ref to pod", "pod", klog.KObj(pod), "containerName", container.Name)
	}

	// If the image contains no tag or digest, a default tag should be applied.
	image, err := applyDefaultImageTag(container.Image)
	if err != nil {
		msg := fmt.Sprintf("Failed to apply default image tag %q: %v", container.Image, err)
		m.logIt(ref, v1.EventTypeWarning, events.FailedToInspectImage, logPrefix, msg, klog.Warning)
		return "", msg, ErrInvalidImageName
	}
```

We call `GenerateContainerRef` and that gives a `ref`.  We then use the `ref` in the `m.logIt`.

Reply:
sorry I dont see how `ref` is related to `applyDefaultImageTag`. 
Are you concerned that moving `if strings.Contains(image, ":latest")` to the beginning of `applyDefaultImageTag` will change the returned error? In other words, an invalid image name with `:latest` would not return an error if we moved `if strings.Contains(image, ":latest")`?

Reply:
I'm saying that we call `GenerateContainerRef` to get a `ref`.  And then we use `ref` in `m.logIt` after we call `applyDefaultImageTag`.

If we fail `applyDefaultImageTag` we should log but we are using `ref` in that log call so not sure if we can move `applyDefaultImageTag` to first call and log without calling `generateContainerRef` also.